arviz/stats/diagnostics.py

```diff
@@ -996,3 +996,134 @@ def _multichain_statistics(ary, focus="mean"):
+        Names of variables to include in the rhat report
+    component : str
+        Select component to power-scale. Valid components are are:
+        - "prior"
+        - "likelihood"
+    dask_kwargs : dict, optional
+        Dask related kwargs passed to :func:`~arviz.wrap_xarray_ufunc`.
+
+    Returns
+    -------
+    xarray.Dataset
+      Returns dataset of power-scaling sensitivity diagnostic values.
+
+    Notes
+    -----
+    The diagnostic is computed by power-scaling the specified component (prior or likelihood)
+    and determining the degree to which the posterior changes. It uses Pareto-smoothed
+    importance sampling to avoid refitting the model.
+
+    References
+    ----------
+    * Kallioinen et al. (2022) see https://arxiv.org/abs/2107.14054
+
+    """
+    dataset = convert_to_dataset(data, group="posterior")
+    var_names = _var_names(var_names, dataset)
+
+    dataset = dataset if var_names is None else dataset[var_names]
+
+    if (component == "likelihood"):
+        component_draws = np.sum(data["log_likelihood"].stack(draws=("chain", "draw")).obs.values, axis=0)
+    elif (component == "prior"):
+        component_draws = data["log_prior"].stack(draws = ("chain", "draw")).values
+
+    lower_w = np.exp(_powerscale_lw(component_draws=component_draws, alpha=alpha))
+    lower_w = lower_w/np.sum(lower_w)
+
+    upper_w = np.exp(_powerscale_lw(component_draws=component_draws, alpha=alpha))
+    upper_w = upper_w/np.sum(upper_w)
+
+    ufunc_kwargs = {"ravel": False}
+    func_kwargs = {
+        "lower_weights": lower_w,
+        "upper_weights": upper_w,
+        "delta": delta
+    }
+
+    return _wrap_xarray_ufunc(
+        _powerscale_sens,
+        dataset,
+        ufunc_kwargs=ufunc_kwargs,
+        func_kwargs=func_kwargs,
+        dask_kwargs=dask_kwargs,
+    )
+
+
+def _powerscale_sens(ary, *, lower_weights=None, upper_weights=None, delta=0.01, dask_kwargs=None):
+    """
+    Calculate power-scaling sensitivity by finite difference second derivative of CJS
+    """
+    lower_cjs = _cjs_dist(x=draws, weights=lower_weights)
+    upper_cjs = _cjs_dist(x=draws, weights=upper_weights)
+
+    logdiffsquare = 2 * np.log2(1 + delta)
+    grad = (lower_cjs + upper_cjs) / logdiffsquare
+
+    return grad
+
+
+def _powerscale_lw(alpha, component_draws):
+    """
+    Calculate log weights for power-scaling component by alpha.
+    """
+    lw = (alpha - 1) * component_draws
+    lw = psislw(lw)
+
+    return lw
+
+
+def _cjs_dist(x, weights):
+    """
+    Calculate Cumulative Jensen-Shannon distance between original draws (x) and weighted draws.
+    """
+
+    # normalise weights
+    weights = weights / np.sum(weights)
+
+    # sort draws and weights
+    x, w = (list(x) for x in zip(*sorted(zip(x, weights))))
+
+    bins = x[:-1]
+    binwidth = np.diff(x)
+
+    # ecdfs
+    cdf_p = np.full(shape=len(x), fill_value=1/len(x))
+    cdf_p = np.cumsum(cdf_p)[:-1]
+    cdf_q = np.cumsum(w/np.sum(w))[:-1]
```

Comment:
would it make a difference to subset `[1:]` instead? That is, get an array starting from the first element and ending at 1. I think now the first element is potentially 0 and can give raise to nans, so doing this would allow us to use sum instead of nansum.

Reply:
> Re storage of the inputs @ahartikainen. I think here what we need are the log_prior and the log_likelihood (not pointwise) of the posterior samples, so it would probably be best to add a `log_prior` and `log_likelihood` variables to `sample_stats` (like we have `lp` already, in fact, the sum of these two should be equal to `lp` so it might make sense to store only the `log_prior` for example and get the log likelihood by substracting it.

In simple cases of power-scaling, the log_prior + log_likelihood = lp, but this is not true in general. For example, in a hierarchical model, we suggest only power-scaling the top-level prior (not intermediate), so this subtraction would not work (see paper for more details). In these cases, the 'log_prior' is not the joint global prior, but is only selected priors (which should somehow be specified by the modeller).

Reply:
let's keep the initial log_likelihood and log_prior proposals then.

is it possible to update the slicing though?

Reply:
> is it possible to update the slicing though?

I think this might change the calculation of the metric as defined in https://link.springer.com/content/pdf/10.1007/978-3-319-23525-7_11.pdf  (Definition 1, and section 2.3).

In the paper 0 * log 0 = 0 is defined (nansum takes care of this). In a quick test, changing the slicing resulted in quite a different output value.
Is it critical that we use np.sum instead of np.nansum? I can spend some more time thinking how to fix the calculation if needed (unless you have an idea already?).